server/interpreter.js

```diff
@@ -2103,42 +2114,78 @@ Interpreter.prototype.throwException = function(value, opt_message) {
     }
     error = new this.Error(value, opt_message);
   }
-  this.executeException(error);
+  this.unwind(Interpreter.Completion.THROW, error, undefined);
   // Abort anything related to the current step.
   throw Interpreter.STEP_ERROR;
 };
 
 /**
- * Throw an exception in the interpreter that can be handled by an
- * interpreter try/catch statement.  If unhandled, a real exception will
- * be thrown.
- * @param {Interpreter.Value} error Value being thrown.
+ * Unwind the stack to the innermost relevant enclosing TryStatement,
+ * For/ForIn/WhileStatement or Call/NewExpression.  If this results in
+ * the stack being completely unwound the thread will be terminated
+ * and an appropriate error being logged.
+ * @param {Interpreter.Completion} type Completion type.
+ * @param {Interpreter.Value=} value Value computed, returned or thrown.
+ * @param {string=} label Target label for break or return.
  */
-Interpreter.prototype.executeException = function(error) {
-  // Search for a try statement.
+Interpreter.prototype.unwind = function(type, value, label) {
+  if (type === Interpreter.Completion.NORMAL) {
+    throw TypeError('Should not unwind for NORMAL completions.');
+  }
+
   for (var stack = this.thread.stateStack; stack.length > 0; stack.pop()) {
     var state = stack[stack.length - 1];
-    if (state.node['type'] === 'TryStatement') {
-      state.throwValue = error;
-      return;
+    switch (state.node['type']) {
+      case 'TryStatement':
+        state.cv = {type: type, value: value, label: label};
+        return;
+      case 'CallExpression':
+      case 'NewExpression':
+        switch (type) {
+          case Interpreter.Completion.BREAK:
+          case Interpreter.Completion.CONTINUE:
+            throw Error('Unsynatctic break/continue not rejected by acorn');
+          case Interpreter.Completion.RETURN:
+            state.value = value;
+            return;
+        }
+        break;
+    }
+    if (type === Interpreter.Completion.BREAK) {
+      if (label === undefined ?
+          (state.isLoop || state.isSwitch) :
+          (state.labels && state.labels.indexOf(label) !== -1)) {
+        stack.pop();
+        return;
+      }
+    } else if (type === Interpreter.Completion.CONTINUE) {
+      if (label == undefined ?
```

Comment:
The spec defines labels as identifiers, so the empty string is not valid.  Label sets are either non-empty collections of labels, or the distinguished value `empty` (not an identifier spelled that way, but a special sentinel value).  It represents unlabelled breaks as "(break, empty, empty)"; I think we could represent them as either `''` (as I did in the Go interpreter) or `undefined` (as I do here) or even `null`.  I don't see any obvious reason to pick one over the others, other than that using `null` or `undefined` seems slightly clearer than just using the empty string.

As to this particular bit of *code*: it's probably best to use `!label` here as it will be most tolerant to incorrect input and is most clear about the particular information we are interested in (is there a label?)